rfcs/0001-meta-schedule-autotensorir.md

```diff
@@ -0,0 +1,444 @@
+
+```python
+# requires to trace the execution
+sch = tir.Schedule(..., traced=True)
+# do a lot of scheduling
+...
+# print the trace
+print(sch.trace)
+```
+
+And below is an example of the printed trace, which honestly reflects the schedule as a snippet of
+python scheduling function:
+
+```python
+b0 = sch.get_block(name="matmul", func_name="main")
+l1, l2, l3 = sch.get_loops(block=b0)
+v4, v5, v6, v7 = sch.sample_perfect_tile(loop=l1, n=4, max_innermost_factor=16, decision=[32, 1, 16, 2])
+v8, v9, v10, v11 = sch.sample_perfect_tile(loop=l2, n=4, max_innermost_factor=16, decision=[64, 4, 2, 2])
+v12, v13 = sch.sample_perfect_tile(loop=l3, n=2, max_innermost_factor=16, decision=[64, 16])
+l14, l15, l16, l17 = sch.split(loop=l1, factors=[v4, v5, v6, v7])
+l18, l19, l20, l21 = sch.split(loop=l2, factors=[v8, v9, v10, v11])
+l22, l23 = sch.split(loop=l3, factors=[v12, v13])
+sch.reorder(l14, l18, l15, l19, l22, l16, l20, l23, l17, l21)
+```
+
+### 4.2. Exploring the Design Space
+
+Meta Schedule provides several built-in exploration strategies to exhaustively or efficiently search
+for efficient schedules.
+
+**Random search by replaying schedule functions.** With a user-provided schedule function
+as a black-box design space generator, our system could repetitively invoke such an opaque function
+without doing any extra analysis. The function could be written in C++ or Python, or any language
+that implements packed function FFI. If sampling instructions are present in the function, then each
+invocation results in a different IRModule after being scheduled because the random decisions are
+possibly changed across different runs. Effectively, it is equivalent to
+random exploration without trace, allowing the flexibility for users to define arbitrary functions
+that trace may not well support (e.g. control flow divergence based on the value of intermediate
+random variables), but it forbids future opportunity of any trace-based analysis.
+
+**Random search by replaying traces.** Traces are obtained from a design space generator, and
+replayed with a builtin interpreter in our system. If sampling instructions are present on the
+traces, then their random decisions are mutated during each replay, i.e. jumps to a new point in the
+design space. Therefore, repetitive replay of those traces are equivalent to exploration of the
+design space. Our system could potentially benefit from trace-based analysis, including rejecting
+obviously invalid schedules (e.g. using too much CUDA resources), doing dead-code elimination to
+simplify a trace, extracting trace-based features used in the cost model, etc.
+
+**Cost-model-guided evolutionary search**. A more efficient exploration strategy. We define two sets
+of rules:
+
+* Mutator: defines how to jump to a point’s "neighbor" in the design space
+* Postprocessor: after the trace is executed, there are some extra rules we want to execute, for
```

Comment:
"rules" sounds like SearchRule. could you say something like "there may be extra rule-specific logic to execute to either validate or tweak the produced schedule."